diff --git a/src/rite/rite.odin b/src/rite/rite.odin
index be41836..037536f 100644
--- a/src/rite/rite.odin
+++ b/src/rite/rite.odin
@@ -1514,6 +1514,54 @@ delete_code_builder :: proc(builder: ^CodeBuilder) {
 // Constants ======================================================================================
 
 const_value :: proc(builder: ^CodeBuilder, value: Value) -> int {
+	for i := 0; i < len(builder.constants); i += 1 {
+		existing := builder.constants[i]
+
+		switch v in value {
+		case bool:
+			existing_bool, is_bool := existing.(bool)
+			if is_bool && existing_bool == v {
+				return i
+			}
+
+		case i64:
+			existing_int, is_int := existing.(i64)
+			if is_int && existing_int == v {
+				return i
+			}
+
+		case f64:
+			existing_float, is_float := existing.(f64)
+			if is_float && existing_float == v {
+				return i
+			}
+
+		case ^Object:
+			existing_object, is_object := existing.(^Object)
+			if !is_object || existing_object.kind != v.kind {
+				continue
+			}
+
+			if v.kind == .STRING {
+				existing_string := cast(^StringObject)existing_object
+				value_string := cast(^StringObject)v
+				if existing_string.text == value_string.text {
+					return i
+				}
+				continue
+			}
+
+			if existing_object == v {
+				return i
+			}
+		}
+	}
+
+	if len(builder.constants) >= int(max(u16)) + 1 {
+		compile_error("code uses too many constants")
+		return 0
+	}
+
 	append(&builder.constants, value)
 	return len(builder.constants) - 1
 }
@@ -1856,12 +1904,9 @@ resolve_upvalue :: proc(builder: ^CodeBuilder, symbol: ^SymbolObject) -> (int, b
 }
 
 compile_constant :: proc(builder: ^CodeBuilder, value: Value, dst: int) {
-	if len(builder.constants) > int(max(u16)) {
-		compile_error("code uses too many constants")
-		return
-	}
-
 	constant_index := const_value(builder, value)
+	if Compiler.failed { return }
+
 	emit_load_const(builder, dst, constant_index)
 }
 
